pkg/apis/pipeline/v1alpha1/pull_request_resource.go

```diff
@@ -0,0 +1,159 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"flag"
+	"fmt"
+	"strings"
+
+	"github.com/tektoncd/pipeline/pkg/names"
+	corev1 "k8s.io/api/core/v1"
+)
+
+var (
+	prSource = "pr-source"
+	// The container that we use to implement the PR source step.
+	prImage = flag.String("pr-image", "override-with-pr:latest",
+		"The container image containing our PR binary.")
+)
+
+// PullRequestResource is an endpoint from which to get data which is required
+// by a Build/Task for context.
+type PullRequestResource struct {
+	Name             string               `json:"name"`
+	Type             PipelineResourceType `json:"type"`
+	URL              string               `json:"url"`
+	GithubOauthToken string               `json:"githubOauthToken"`
+	//Secrets holds a struct to indicate a field name and corresponding secret name to populate it
+	Secrets []SecretParam `json:"secrets"`
+}
+
+// NewPullRequestResource create a new git resource to pass to a Task
+func NewPullRequestResource(r *PipelineResource) (*PullRequestResource, error) {
+	if r.Spec.Type != PipelineResourceTypePullRequest {
+		return nil, fmt.Errorf("PipelineResource: Cannot create a PR resource from a %s Pipeline Resource", r.Spec.Type)
+	}
+	prResource := PullRequestResource{
+		Name:    r.Name,
+		Type:    r.Spec.Type,
+		Secrets: r.Spec.SecretParams,
+	}
+	for _, param := range r.Spec.Params {
+		switch {
+		case strings.EqualFold(param.Name, "URL"):
+			prResource.URL = param.Value
+		}
+	}
+
+	return &prResource, nil
+}
+
+// GetName returns the name of the resource
+func (s PullRequestResource) GetName() string {
+	return s.Name
+}
+
+// GetType returns the type of the resource, in this case "Git"
+func (s PullRequestResource) GetType() PipelineResourceType {
+	return PipelineResourceTypePullRequest
+}
+
+// GetURL returns the url to be used with this resource
+func (s *PullRequestResource) GetURL() string {
+	return s.URL
+}
+
+// GetParams returns the resource params
+func (s PullRequestResource) GetParams() []Param { return []Param{} }
+
+// Replacements is used for template replacement on a PullRequestResource inside of a Taskrun.
+func (s *PullRequestResource) Replacements() map[string]string {
+	return map[string]string{
+		"name": s.Name,
+		"type": string(s.Type),
+		"url":  s.URL,
+	}
+}
+
+func (s *PullRequestResource) GetDownloadContainerSpec() ([]corev1.Container, error) {
+	args := []string{"-url", s.URL, "-path", s.Name, "-mode", "download"}
+
+	evs := []corev1.EnvVar{}
+	for _, sec := range s.Secrets {
+		switch {
+		case strings.EqualFold(sec.FieldName, "githubOauthToken"):
```

Comment:
why are secrets only added if they have this particular name? (maybe im misunderstanding!)

Reply:
The thought was to pull the value of the token from any (secret name, secret key). We use the FieldName to denote which secret should be used for this so that you can define your k8s with whatever name/key you want and set it in the ResourceSpec with the specified name.

No idea if this is the right way to go about doing this. If there is a better way somewhere else let me know!

Added this to the documentation.